prereise/gather/hiflddata/data_trans.py

```diff
@@ -0,0 +1,645 @@
+            continue
+
+        start_coord = (
+            line["geometry"]["coordinates"][0][0][1],
+            line["geometry"]["coordinates"][0][0][0],
+        )
+        end_coord = (
+            line["geometry"]["coordinates"][0][-1][1],
+            line["geometry"]["coordinates"][0][-1][0],
+        )
+        candidate_coordinates = sub_name_dict.get(
+            line["properties"]["SUB_1"]
+        ) + sub_name_dict.get(line["properties"]["SUB_2"])
+
+        sub1 = min(candidate_coordinates, key=lambda p: computeGeoDist(p, start_coord))
+        sub2 = min(candidate_coordinates, key=lambda p: computeGeoDist(p, end_coord))
+
+        dist = computeGeoDist(start_coord, end_coord)
+
+        lines.append(
+            [
+                line_id,
+                line_type,
+                sub_by_coord_dict.get(sub1)[0],
+                sub_by_coord_dict.get(sub1)[1],
+                sub_by_coord_dict.get(sub2)[0],
+                sub_by_coord_dict.get(sub2)[1],
+                dist,
+            ]
+        )
+
+        if sub1 == sub2:
+            continue
+        if sub1 in N_dict:
+            N_dict[sub1].append(sub2)
+        else:
+            nodes.append(sub1)
+            N_dict[sub1] = [sub2]
+        if sub2 in N_dict:
+            N_dict[sub2].append(sub1)
+        else:
+            nodes.append(sub2)
+            N_dict[sub2] = [sub1]
+    return lines, nodes, N_dict
+
+
+def lineFromCSV(T_csv):
+    """Create dict to store all the raw transmission line csv data
+
+    :param str T_csv: path of the HIFLD transmission csv file
+    :return: (*dict*) -- a dict mapping the transmission ID to its raw parameters.
+    """
+
+    raw_data = pd.read_csv(T_csv)
+    raw_data["ID"] = raw_data["ID"].astype("str")
+    raw_lines = raw_data.set_index("ID").to_dict()
+    return raw_lines
+
+
+def meter2Mile(dist):
+    """Calculate the mile given the distance in meter
+
+    :param float dist: length of the distance between two substations
+    :return: (*float*) -- distance in mile unit.
+    """
+
+    return dist / 1609.34
+
+
+def computeGeoDist(sub1, sub2):
+    """Compute the haversine geo distance of two substations
+
+    :param tuple sub1: (latitude, longitude) of substation1
+    :param tuple sub2: (latitude, longitude) of substation2
+    :return: (*float*) -- distance between two geo coordinates.
+    """
+
+    return haversine(sub1, sub2, Unit.MILES)
+
+
+def GraphOfNet(nodes, N_dict):
+    """Create the graph for the HIFLD transmission network
+
+    :param list nodes: list of substation as the vertex inside the Graph
+    :param dict N_dict: dict of substation's neighbors as the edge inside the Graph
+    :return: (*networkx.Graph*) -- Graph generated by networkx.
+    """
+
+    G = nx.Graph()
+    for node in nodes:
+        G.add_node(node)
+
+    for key in N_dict:
+        for value in N_dict[key]:
+            G.add_edge(key, value)
+
+    return G
+
+
+def GetMaxIsland(G):
```

Comment:
Can this function be replaced with `list(max(nx.connected_components(G), key=len))`?

Reply:
Awesome! this will get exactly the same result. will replace in next revision. thanks.